pkg/morph/client/notifications.go

```diff
@@ -211,6 +211,29 @@ func (c *Client) restoreSubscriptions(cli *rpcclient.WSClient, endpoint string)
 		id  string
 	)
 
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	// neo-go WS client says to _always_ read notifications
+	// from its channel. Subscribing to any notification
+	// while not reading them in another goroutine may
+	// lead to a dead-lock, thus that async side notification
+	// listening while restoring subscriptions
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				return
+			case n, ok := <-cli.Notifications:
+				if !ok {
+					return
+				}
+
+				c.notifications <- n
```

Comment:
Is it possible that the reader of `c.notifications` will be blocked by some task and stop reading?

Reply:
It is possible during shutdown, but we already have the same problem in the main client.
It will be fixed in the main branch (or will go away together with refactoring and adopting the not-so-new actor API here :))

Reply:
> Is it possible that the reader (dispatcher) of c.notifications will be blocked by some task and stop reading?

it could happen in theory but we did all we could (handling via async pool): https://github.com/nspcc-dev/neofs-node/blob/d4b108d8065f8268a988cf3966e181f02e1e5af7/pkg/morph/event/listener.go#L254-L299

do not see any problem with that: a new routine in that PR does the same thing the main notiication routine usually does (but not in `restoreSubscriptions()` since it just could not do that bacause it is busy with the switching process). so it is just a temporary notification listening. it blocks the same way the main `notificationLoop` does.

> You might as well split reconnection and resubscription.

@roman-khimov, sure. that PR is ugly but it works. we just cannot afford to make any huge changes in the `support` branches. both `frostfs-node` and `neofs-node` will change that code soon anyway (new notification API has to be adopted)